Add helm chart for unitycatalog

sdwbgn opened the pull request:
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add tests! [not applicable]
- [x] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

Bringing helm chart from - https://github.com/sdwbgn/unitycatalog-helm

Features:
- Deploys UC Server and UI
- UC Configuration from helm values
  - OAuth
  - Data sources configuration (S3, GCS, ADLS) with secret management
- Persistence of data in PVCs

Relates: https://github.com/unitycatalog/unitycatalog/issues/433

tnightengale commented:
@sdwbgn This is really exciting! Currently #698 only pushes the backend image. Is it your recommendation for hosting on k8s that we also create and push a `unitycatalog-ui` image?

tnightengale commented:
My instinct is that we should overhaul the current `ui/Dockerfile` to be production quality, build the react app, and then serve it using nginx in a second stage of a multistage Dockerfile. What are your thoughts @sdwbgn @jamieknight-db?

sdwbgn commented:
> Is it your recommendation for hosting on k8s that we also create and push a unitycatalog-ui image?
> Or for example, would it be better to bundle the ui into the backend image?

Oh, I thought both images will be published in that PR. Usually it is best practice to have images as decoupled as possible. IMO best approach would be to separate image for backend and frontend.

> then serve it using nginx in a second stage of a multistage Dockerfile

In K8s deploying the app and serving it are usually separate. It's mostly common for Helm charts to deploy bare minimum - the app itself and service for accessing the app.
After that, for example, one of the options is to expose your service by changing Service type to NodePort (to expose raw port on Kubernetes compute node) or LoadBalancer (to use your cloud provider's solution for ALB/NLB).
I think we could also integrate add Ingress to the chart to use it with controllers, like nginx-ingress.

tnightengale commented:
> #902

@dennyglee I start a new role that week, so unclear if I will have a conflict. But will endeavor to attend!

sdwbgn commented:
> I thought it would be preferable to support TLS.

Makes sense! Currently we can achieve that by terminating TLS/SSL connection on NLB or other ingress. Additionally we might want to consider support for TLS in UC itself.

dennyglee commented:
Adding @roeap to review!

roeap reviewed (commented):
Overall I think the Helm chart looks good, however I do have some questions around how we manage storage - and please excuse if I am completely off track here :)

IIRC, in a StatefulSet, each Pod is assigned dedicated storage (PV & PVC) isolated from all other pods. So when using it for the file based storage, we may end up with each replica effectively having its own version of the on-disk database that would be always out of sync?

For the configuration - since we render it on ever pod-start - i think we can just use an ephemeral volume and avoid creating a PV?

If the above is infact true, we may wnat to consider using a Deployment and creating a separate `RedWritemany` PV/PVC that all pods can mount, or force all pods on the same node via affinities somehow? Or, since on-dosk storage seems most suited for evaluation deployments, just limit it to ine replica in that case? This would then allow a wider range of storages, since I believe ony a limited sub-set supports RWMany ..

I was a little suprised by the storage credentials, but I guess this is simply due to the fact, that the `Credentials` api surface is not implemented yet, so there is no dynamic handling of credentials possible?

sdwbgn commented on helm/chart/templates/server/statefulset.yaml:
```diff
@@ -0,0 +1,222 @@
+        - name: config-volume
+          emptyDir: {}
+      {{- end }}
+      {{- if not (and (eq .Values.db.type "file") .Values.db.fileConfig.persistence.enabled) }}
+        - name: db-volume
+          emptyDir: {}
+      {{- end }}
```
Good point, will fix that!
I will keep db volume mounts for "file" type

sdwbgn replied:
Added postgresql DB support, fixed it

sdwbgn commented on helm/chart/templates/server/statefulset.yaml:
```diff
@@ -0,0 +1,222 @@
+            - name: api
+              containerPort: {{ .Values.server.statefulset.port }}
+              protocol: TCP
+          {{- with .Values.server.statefulset.livenessProbe }}
+          livenessProbe:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
```
Agreed, will remove default livenessProbe

sdwbgn replied:
Done

captainnemo27 commented:
i think cannot got to prodution :< , i am trying but get a lot of missing config

hongbo-miao commented on helm/chart/values.yaml:
```diff
@@ -0,0 +1,190 @@
+db:
+  type: file
+  # Supported values: file
+  # TODO: add support for postgres and mysql
```
Thanks for this pull request ❤️ Cannot wait!
I assume in future this Postgres will help avoid using Persistent Volume, then can support multiple replicas (?)

sdwbgn replied:
I added postgresql support, tested simple case of multi-replica UC (without auth) and it seems to work

ognis1205 commented:
Any updates on this?

dennyglee commented:
@roeap Can you review when you get the chance?  Thanks!

sdwbgn commented:
Sorry for long wait, I am currently working on current comments, I will push new revision today or tomorrow.

sdwbgn commented on helm/chart/templates/server/statefulset.yaml:
```diff
@@ -0,0 +1,222 @@
+    - metadata:
+        name: db-volume
+      spec:
+        accessModes: {{ .Values.db.fileConfig.persistence.accessModes }}
```
I would consider file-based DB to be single replica solution, hence default value is `[ "ReadWriteOnce" ]`. ReadWriteMany implementations in most clouds are not perfectly reliable, have quite a lot of limitations (performance, latency, race conditions) and overall cover very specific use cases.

Setting multiple accessModes is incredibly rare case, I never seen a case in production where that would be a requirement. If that is really needed, it currently can be overridden in values.yaml, but as for the default value ReadWriteOnce should be enough.

sdwbgn commented on helm/chart/templates/server/statefulset.yaml:
```diff
@@ -0,0 +1,222 @@
+          requests:
+            storage: {{ .Values.server.config.persistence.size }}
+        storageClassName: {{ .Values.server.config.persistence.storageClassName }}
+    {{- end }}
```
There was a need for that due to JWT private key creation by UC. I moved it to Secret resource and removed the need for config persistence (but I need to test it further).

sdwbgn commented on helm/chart/templates/_helpers.tpl:
```diff
@@ -0,0 +1,114 @@
+*/}}
+{{- define "unitycatalog.server.apiEndpoint" -}}
+http://{{ include "unitycatalog.server.fullname" . }}:{{ .Values.server.service.port }}
+{{- end }}
```
Done

sdwbgn commented:
Published next iteration
- Added HTTPRoute and Ingress support
  - UI/API requests routing on same hosts based on prefix (/api prefix for API, everything else to UI pods)
  - Allows for disabling proxy for API requests in UI (required for using Ingress / HTTPRoute)
- Added PostgreSQL support
  - Allows for multi-replica deployments
- Refactored config creation to remove need for persisting config values in separate volume
- Generated docs using `helm-docs`
  - Added precommit check to keep it in-sync
- Using deployment instead of statefulset

Currently, I have tested following:
- Set up UC with PostgreSQL and 2 replicas for both UI and API
- Set up HTTPRoute with public DNS record and exposed to internet
  - Used Envoy Gateway as Gateway API provider
  - Restricted access using EnvoySecurityPolicy (basic-auth)
- Set up SSL certificate using cert-manager and referenced in helm chart

What's left to test:
- Test S3/ADLS/GCS credential management
- External auth - could be broken during refactoring of secrets and keys

Overall there is still need for proper UI docker image.

sdwbgn commented on helm/values.httproute.yaml:
Example of values.yaml for prod-like deployment with 2 replicas, PostgreSQL and routing using HTTPRoute

unfrgivn commented:
@sdwbgn Thanks for pushing on this and also adding Postgres support. Any interest in adding the typical array config for `extraContainers` inside the Deployment template so sidecars like SQL proxy can be added?

unfrgivn commented:
@sdwbgn A couple items on the Gateway implementation. I didn't dig into that template specifically since I wouldn't use that and I don't think the Gateway object belongs in the chart. My simple proposal would be to change the default `enabled: false` when enabling the HTTPRoute. The Gateway API is great and I hope support for it starts appearing in more charts, but there's too many different implementations and dependencies on the cluster setup and CSP that will change how it's implemented. 

Typically I add my routes and backendpolicy k8s objects via Kustomize on top of Helm. So if I were to use this I would disable the Gateway/HTTPRoute in the chart, but then it will try to use the frontend proxy in the Deployment. My second choice would be to enable the HTTPRoute using the `extraParentRefs`, disable the Gateway creation, and use Kustomize to patch the hosts and add another route to enforce HTTPS redirect.

And a minor correction, the chart docs say that `httpRoute.host` is an array but it's implemented in the template as a string value that gets added to the `spec.hostnames` array. 

I am happy to collaborate more on this if you would like, but I'm also not trying to throw a bunch of changes into this PR as I know the goal is to get a first iteration merged.

sdwbgn commented:
@unfrgivn Thanks for your comments!
> Any interest in adding the typical array config for extraContainers inside the Deployment template so sidecars like SQL proxy can be added?

Nice idea, added.

> I didn't dig into that template specifically since I wouldn't use that and I don't think the Gateway object belongs in the chart.

I gave it a thought, and I agree with you, removed it and left only `parentRefs`

> And a minor correction, the chart docs say that httpRoute.host is an array but it's implemented in the template as a string value that gets added to the spec.hostnames array.

Fixed the doc type, thanks!

roeap reviewed (approved):
@sdwbgn - looking great!

Since this is quite a complex thing with many nuances I expect we may need to iterate once this gets into the wild, but this is expected :).

One more update from main an need to fix the DCO, gut I think good to go.

Gerrit-K commented on helm/values.yaml:
```diff
@@ -0,0 +1,862 @@
+      #
+      # [Kubernetes docs](https://kubernetes.io/docs/concepts/policy/security-context/)
+      #
+      securityContext:
```
As a general suggestion, I would propose to add a few settings by default and let users decide whether they need to extend/override the securityContext. A sensible default could be:
```yaml
podSecurityContext:
  runAsNonRoot: true
  runAsUser: 100
  runAsGroup: 101
  fsGroup: 101
  seccompProfile: { type: RuntimeDefault }

securityContext:
  allowPrivilegeEscalation: false
  capabilities: { drop: [ALL] }
```
At least the `runAs...` settings are useful so that the users don't need to inspect the image to get the user and group IDs to run the container with.

sdwbgn replied:
Great idea! Currently, due to the alpine images and package installation, it's not possible to drop root permissions. I have prepared a PR with adding those packages to main UC image - https://github.com/unitycatalog/unitycatalog/pull/1019. After it's merged, I will drop bare alpine image and usage of `apk`, and we can use these defaults.

Gerrit-K replied:
Thanks for the quick reply! Sounds reasonable 👍 
Looking forward to the further improvements

Gerrit-K reviewed (commented):
Hey @sdwbgn sorry for randomly chiming in here with a review. I was evaluating unity with my team and we were happy to see that there might be an official helm chart just around the corner. Unfortunately, while trying it out, we stumbled over a few things that we couldn't work around - most of which was related to security enforcements of our target cluster. I figured I might comment some of them here already, but please don't see this as a blocker for the PR. 
Thanks for your work!

sdwbgn commented on helm/values.yaml:
```diff
@@ -0,0 +1,862 @@
+    # The secret must contain the following keys:
+    # - `public_key.der`: Public key in DER format
+    # - `private_key.der`: Private key in DER format
+    # - `key_id.txt`: Key ID
```
Thanks for pointing it out! Fixed

sdwbgn commented on helm/templates/server/deployment.yaml:
```diff
@@ -0,0 +1,249 @@
+            - /bin/sh
+            - -c
+            - |-
+              apk add --no-cache envsubst openssl curl
```
Made a separate PR with that - https://github.com/unitycatalog/unitycatalog/pull/1019. Will update the chart as soon as we merge it.

sdwbgn commented:
@Gerrit-K Thanks for your suggestions! They are exactly what was missing for making the deployment more secure. I've addressed the certificate issue and securityContext for initContainer of UI. For the init image and default securityContext I want to make a separate change, after PR https://github.com/unitycatalog/unitycatalog/pull/1019 is merged.
